## Uploading IRFinder output files into R
## pathMONvsMAC.txt contains directories where the output files are stored for each biological replicate

## importing ensembl features
library(biomaRt)
grch38<-useMart("ensembl",dataset="hsapiens_gene_ensembl")
database<-getBM(attributes = c("ensembl_gene_id", "ensembl_gene_id_version", "ensembl_transcript_id", "ensembl_transcript_id_version", "percentage_gene_gc_content", "go_id", "name_1006", "ccds"), filters="ensembl_gene_id", values= IRdata[[1]]$EnsemblID, mart=grch38)
TSS<-read.delim("hg38_TSS.txt", header = T)

files<-read.table(file = "pathsMONvsMAC.txt", header = F)
files = as.vector(files$V1)

for (i in 1:length(files)) {
  IRdata[[i]] <- read.table(files[i], header = F)
  colnames(IRdata[[i]])<-c("Chromosome", "Start", "End", "Gene Symbol/Ensembl ID/Static Warning", "Empty", "Strand", "ExclBases", "Coverage", "IntronDepth", "Intron25Percentile", "Intron50Percentile", "Intron75Percentile", "ExonToIntronReadsLeft", "ExonToIntronReadsRight", "IntronDepthFirst50bp", "IntronDepthLast50bp", "SplicesLeft", "SplicesRight", "SplicesExact", "IRratio", "Warnings")
  IRdata[[i]]$Chromosome<-sub("^", "chr", IRdata[[i]]$Chromosome)
  IRdata[[i]]$Chromosome<-factor(IRdata[[i]]$Chromosome, levels=chrOrder)
  IRdata[[i]]<-IRdata[[i]][order(IRdata[[i]]$Chromosome),]
  IRdata[[i]]$Length<-IRdata[[i]]$End - IRdata[[i]]$Start +1
  IRdata[[i]]$gene<-strsplit(as.character(IRdata[[i]]$`Gene Symbol/Ensembl ID/Static Warning`), split="/")
  IRdata[[i]]$GeneSymbol<-sapply(IRdata[[i]]$gene, "[[", 1)
  IRdata[[i]]$EnsemblID<-sapply(IRdata[[i]]$gene, "[[", 2)
  IRdata[[i]]$StaticWarning<-sapply(IRdata[[i]]$gene, "[[", 3)
  IRdata[[i]]$IntronID<-paste0(IRdata[[i]]$Chromosome,":",IRdata[[i]]$Start,":", IRdata[[i]]$End)
  IRdata[[i]]$ENST<-all.introns$V8[match(paste(IRdata[[i]]$Chromosome, IRdata[[i]]$Start, IRdata[[i]]$End, sep = ":"), paste(all.introns$V1, all.introns$V2, all.introns$V3, sep=":"))]
  IRdata[[i]]$GCcont_gene<-database$percentage_gene_gc_content[match(IRdata[[i]]$ENST, database$ensembl_transcript_id)]
  IRdata[[i]]$GO_id<-database$go_id[match(IRdata[[i]]$ENST, database$ensembl_transcript_id)]
  IRdata[[i]]$GO_term<-database$name_1006[match(IRdata[[i]]$ENST, database$ensembl_transcript_id)]
  IRdata[[i]]$CCDS<-database$ccds[match(IRdata[[i]]$ENST, database$ensembl_transcript_id)]
  IRdata[[i]]$TSS.start<-TSS$Transcript.start..bp.[match(IRdata[[i]]$ENST, TSS$Transcript.stable.ID)]
  IRdata[[i]]$TSS.end<-TSS$Transcript.end..bp.[match(IRdata[[i]]$ENST, TSS$Transcript.stable.ID)]
  IRdata[[i]]$TSS.size<-IRdata[[i]]$TSS.end - IRdata[[i]]$TSS.start + 1
  IRdata[[i]]$trans_type<-TSS$Transcript.type[match(IRdata[[i]]$ENST, TSS$Transcript.stable.ID)]
}

warnings.qc<-list() #removing all rows that have any warning messages, other than "-"##
for (i in 1:length(IRdata)) {
  warnings.qc[[i]]<-with(IRdata[[i]], (Warnings=="-"))
}


##seprating introns into two groups: retained and non-retained

introns.ret.samp<-list() ##filtering introns with IR ratio greater than 0.1##
for (i in 1:5) {
  introns.ret.samp[[i]]<-subset((IRdata[[i]][warnings.qc[[i]],]), IRratio>=0.1)
  introns.ret.samp[[i]]$Middle<-round((introns.ret.samp[[i]]$Start + introns.ret.samp[[i]]$End)/2) + 1
  introns.ret.samp[[i]]$Dist.TSS<-introns.ret.samp[[i]]$Start - introns.ret.samp[[i]]$TSS.start
  introns.ret.samp[[i]]$Check<-introns.ret.samp[[i]]$End<introns.ret.samp[[i]]$TSS.end
}


## identifying intron position relative to TSS
for (i in 1:length(introns.ret.samp)) {
  introns.ret.samp[[i]]$Proxim<-sapply(introns.ret.samp[[i]]$Dist.TSS/introns.ret.samp[[i]]$TSS.size, function(x, na.rm = T) if (is.na(x)) "NA"
                                       else if (x<=0.33) "5' site"
                                       else if (x<=0.66) "middle"
                                       else if (x<=1.1) "3' site")}

## using introns.unique.bed generated by IRFinder, determining First and Last introns
all.introns<-read.delim("introns.unique.bed", header = F) #This file is created using a modified perl script
all.introns$V1<-sub("^", "chr", all.introns$V1)
all.introns$V7<-strsplit(as.character(all.introns$V4), split="/")
all.introns$V8<-sapply(all.introns$V7, "[[", 1)
all.introns$V9<-sapply(all.introns$V7, "[[", 2)
all.introns$V10<-sapply(all.introns$V7, "[[", 3)
all.introns$V11<-paste0(all.introns$V1,":",all.introns$V2,":", all.introns$V3)
all.introns<-all.introns[order(all.introns$V8, all.introns$V10),]
all.introns.pos<-all.introns[all.introns$V6 == "+",]
all.introns.neg<-all.introns[all.introns$V6 == "-",]
first.intron.pos<-all.introns.pos[match(unique(all.introns.pos$V8), all.introns.pos$V8),]
first.intron.neg<-all.introns.neg[match(unique(all.introns.neg$V8), all.introns.neg$V8)-1,]
last.intron.pos<-all.introns.pos[match(unique(all.introns.pos$V8), all.introns.pos$V8)-1,]
last.intron.neg<-all.introns.neg[match(unique(all.introns.neg$V8), all.introns.neg$V8),]
last.intron<-rbind(last.intron.pos, last.intron.neg)
first.intron<-rbind(first.intron.pos, first.intron.neg)

introns.ret.samp[[1]]$Proxim[introns.ret.samp[[1]]$IntronID %in% first.intron$V11] <- paste(c("First"))
introns.ret.samp[[1]]$Proxim[introns.ret.samp[[1]]$IntronID %in% last.intron$V11] <- paste(c("Last"))
introns.ret.samp[[2]]$Proxim[introns.ret.samp[[2]]$IntronID %in% first.intron$V11] <- paste(c("First"))
introns.ret.samp[[2]]$Proxim[introns.ret.samp[[2]]$IntronID %in% last.intron$V11] <- paste(c("Last"))
introns.ret.samp[[3]]$Proxim[introns.ret.samp[[3]]$IntronID %in% first.intron$V11] <- paste(c("First"))
introns.ret.samp[[3]]$Proxim[introns.ret.samp[[3]]$IntronID %in% last.intron$V11] <- paste(c("Last"))
introns.ret.samp[[4]]$Proxim[introns.ret.samp[[4]]$IntronID %in% first.intron$V11] <- paste(c("First"))
introns.ret.samp[[4]]$Proxim[introns.ret.samp[[4]]$IntronID %in% last.intron$V11] <- paste(c("Last"))
introns.ret.samp[[5]]$Proxim[introns.ret.samp[[5]]$IntronID %in% first.intron$V11] <- paste(c("First"))
introns.ret.samp[[5]]$Proxim[introns.ret.samp[[5]]$IntronID %in% last.intron$V11] <- paste(c("Last"))


## adding host gene FPKM values
genes.FPKM.files<-c("Gene_expression/BlMo_Hm01_genes.fpkm",
                    "Gene_expression/BlMo_Hm03_genes.fpkm",
                    "Gene_expression/BlMo_Hm05_genes.fpkm",
                    "Gene_expression/BlMa_Hm03_genes.fpkm",
                    "Gene_expression/BlMa_Hm05_genes.fpkm")

genes.FPKM<-list()
for (i in 1:length(introns.ret.samp)) {
  genes.FPKM[[i]]<-read.table(genes.FPKM.files[i], header = T)
  genes.FPKM[[i]][,c(14:19)]<-str_split(as.character(genes.FPKM[[i]]$gene_short_name), ",", 6, simplify = TRUE)
  colnames(genes.FPKM[[i]])[14:19]<-c("gene1", "gene2", "gene3", "gene4", "gene5", "gene6")
}

for (i in 1:length(introns.ret.samp)) {
  introns.ret.samp[[i]]$gene_FPKM<-genes.FPKM[[i]]$FPKM[coalesce(match(introns.ret.samp[[i]]$GeneSymbol, genes.FPKM[[i]][,14]),
                                                                 match(introns.ret.samp[[i]]$GeneSymbol, genes.FPKM[[i]][,15]),
                                                                 match(introns.ret.samp[[i]]$GeneSymbol, genes.FPKM[[i]][,16]),
                                                                 match(introns.ret.samp[[i]]$GeneSymbol, genes.FPKM[[i]][,17]),
                                                                 match(introns.ret.samp[[i]]$GeneSymbol, genes.FPKM[[i]][,18]))]
}

save(introns.ret.samp, file = "introns.ret.samp.RData")

## non-retained introns are defined as below and features are added in the same manner as for the retained introns
introns.nonret.samp<-list()
for (i in 1:5) {
  introns.nonret.samp[[i]]<-subset((IRdata[[i]][warnings.qc[[i]],]), IRratio<=0.01)
  introns.nonret.samp[[i]]$Middle<-round((introns.nonret.samp[[i]]$Start + introns.nonret.samp[[i]]$End)/2) + 1
  introns.nonret.samp[[i]]$Dist.TSS<-introns.nonret.samp[[i]]$Start - introns.nonret.samp[[i]]$TSS.start
  introns.nonret.samp[[i]]$Check<-introns.nonret.samp[[i]]$End<introns.nonret.samp[[i]]$TSS.end
}
