## generating intrinsic features for the introns of interest

## GC content is calculated using bedtools nuc command and intronic coordinates in .bed format
## bedtools output is loaded into R

load("/home/veronikap/MONvsMAC/introns.ret.samp.RData")
BlMo_Hm01.retint.GCcont<-read.delim("/spectrum/GSCT/veronikap/DEEP/GCcount/BlMo/GC_BlMo_Hm01_retint.txt", sep = "\t", header = T)
introns.ret.samp[[1]]$GCcont_int<-BlMo_Hm01.retint.GCcont$X8_pct_gc[match(paste(introns.ret.samp[[1]]$Chromosome,introns.ret.samp[[1]]$Start, sep = ":"), paste0("chr",BlMo_Hm01.retint.GCcont$X.1_usercol,":",BlMo_Hm01.retint.GCcont$X2_usercol))]
BlMo_Hm01.nonretint.GCcont<-read.delim("/spectrum/GSCT/veronikap/DEEP/GCcount/BlMo/GC_BlMo_Hm01_nonretint.txt", sep = "\t", header = T)
introns.nonret.samp[[1]]$GCcont_int<-BlMo_Hm01.nonretint.GCcont$X8_pct_gc[match(paste(introns.nonret.samp[[1]]$Chromosome,introns.nonret.samp[[1]]$Start, sep = ":"), paste0("chr",BlMo_Hm01.nonretint.GCcont$X.1_usercol,":",BlMo_Hm01.nonretint.GCcont$X2_usercol))]

BlMo_Hm03.retint.GCcont<-read.delim("/spectrum/GSCT/veronikap/DEEP/GCcount/BlMo/GC_BlMo_Hm03_retint.txt", sep = "\t", header = T)
introns.ret.samp[[2]]$GCcont_int<-BlMo_Hm03.retint.GCcont$X8_pct_gc[match(paste(introns.ret.samp[[2]]$Chromosome,introns.ret.samp[[2]]$Start, sep = ":"), paste0("chr",BlMo_Hm03.retint.GCcont$X.1_usercol,":",BlMo_Hm03.retint.GCcont$X2_usercol))]
BlMo_Hm03.nonretint.GCcont<-read.delim("/spectrum/GSCT/veronikap/DEEP/GCcount/BlMo/GC_BlMo_Hm03_nonretint.txt", sep = "\t", header = T)
introns.nonret.samp[[2]]$GCcont_int<-BlMo_Hm03.nonretint.GCcont$X8_pct_gc[match(paste(introns.nonret.samp[[2]]$Chromosome,introns.nonret.samp[[2]]$Start, sep = ":"), paste0("chr",BlMo_Hm03.nonretint.GCcont$X.1_usercol,":",BlMo_Hm03.nonretint.GCcont$X2_usercol))]

BlMo_Hm05.retint.GCcont<-read.delim("/spectrum/GSCT/veronikap/DEEP/GCcount/BlMo/GC_BlMo_Hm05_retint.txt", sep = "\t", header = T)
introns.ret.samp[[3]]$GCcont_int<-BlMo_Hm05.retint.GCcont$X8_pct_gc[match(paste(introns.ret.samp[[3]]$Chromosome,introns.ret.samp[[3]]$Start, sep = ":"), paste0("chr",BlMo_Hm05.retint.GCcont$X.1_usercol,":",BlMo_Hm05.retint.GCcont$X2_usercol))]
BlMo_Hm05.nonretint.GCcont<-read.delim("/spectrum/GSCT/veronikap/DEEP/GCcount/BlMo/GC_BlMo_Hm05_nonretint.txt", sep = "\t", header = T)
introns.nonret.samp[[3]]$GCcont_int<-BlMo_Hm05.nonretint.GCcont$X8_pct_gc[match(paste(introns.nonret.samp[[3]]$Chromosome,introns.nonret.samp[[3]]$Start, sep = ":"), paste0("chr",BlMo_Hm05.nonretint.GCcont$X.1_usercol,":",BlMo_Hm05.nonretint.GCcont$X2_usercol))]

BlMa_Hm03.retint.GCcont<-read.delim("/spectrum/GSCT/veronikap/DEEP/GCcount/BlMa/GC_BlMa_Hm03_retint.txt", sep = "\t", header = T)
introns.ret.samp[[4]]$GCcont_int<-BlMa_Hm03.retint.GCcont$X8_pct_gc[match(paste(introns.ret.samp[[4]]$Chromosome,introns.ret.samp[[4]]$Start, sep = ":"), paste0("chr",BlMa_Hm03.retint.GCcont$X.1_usercol,":",BlMa_Hm03.retint.GCcont$X2_usercol))]
BlMa_Hm03.nonretint.GCcont<-read.delim("/spectrum/GSCT/veronikap/DEEP/GCcount/BlMa/GC_BlMa_Hm03_nonretint.txt", sep = "\t", header = T)
introns.nonret.samp[[4]]$GCcont_int<-BlMa_Hm03.nonretint.GCcont$X8_pct_gc[match(paste(introns.nonret.samp[[4]]$Chromosome,introns.nonret.samp[[4]]$Start, sep = ":"), paste0("chr",BlMa_Hm03.nonretint.GCcont$X.1_usercol,":",BlMa_Hm03.nonretint.GCcont$X2_usercol))]

BlMa_Hm05.retint.GCcont<-read.delim("/spectrum/GSCT/veronikap/DEEP/GCcount/BlMa/GC_BlMa_Hm05_retint.txt", sep = "\t", header = T)
introns.ret.samp[[5]]$GCcont_int<-BlMa_Hm05.retint.GCcont$X8_pct_gc[match(paste(introns.ret.samp[[5]]$Chromosome,introns.ret.samp[[5]]$Start, sep = ":"), paste0("chr",BlMa_Hm05.retint.GCcont$X.1_usercol,":",BlMa_Hm05.retint.GCcont$X2_usercol))]
BlMa_Hm05.nonretint.GCcont<-read.delim("/spectrum/GSCT/veronikap/DEEP/GCcount/BlMa/GC_BlMa_Hm05_nonretint.txt", sep = "\t", header = T)
introns.nonret.samp[[5]]$GCcont_int<-BlMa_Hm05.nonretint.GCcont$X8_pct_gc[match(paste(introns.nonret.samp[[5]]$Chromosome,introns.nonret.samp[[5]]$Start, sep = ":"), paste0("chr",BlMa_Hm05.nonretint.GCcont$X.1_usercol,":",BlMa_Hm05.nonretint.GCcont$X2_usercol))]



## flanking exons coordinates are identified using candidate.exons.txt generated by IRFinderPooled

all.exons<-read.delim("/spectrum/GSCT/veronikap/REF/candidate.exons.txt", header = F)
colnames(all.exons)<-c("Chromosome", "Start", "End", "Code", "Empty", "Strand")
all.exons$Chromosome<-sub("^", "chr", all.exons$Chromosome)
all.exons$Code<-strsplit(as.character(all.exons$Code), split="/")
all.exons$ENST<-sapply(all.exons$Code, "[[", 1)


exons.retint.flank5ss<-list()
for (i in 1:length(introns.ret.samp)) {
  exons.retint.flank5ss[[i]]<-all.exons[ifelse(introns.ret.samp[[i]]$Strand=="+", match(paste(introns.ret.samp[[i]]$ENST, introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start, sep = ":"), paste(all.exons$ENST, all.exons$Chromosome, all.exons$End, sep = ":")),
                                               match(paste(introns.ret.samp[[i]]$ENST, introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$End, sep = ":"), paste(all.exons$ENST, all.exons$Chromosome, all.exons$Start, sep = ":"))),]
}

exons.retint.flank3ss<-list()
for (i in 1:length(introns.ret.samp)) {
  exons.retint.flank3ss[[i]]<-all.exons[ifelse(introns.ret.samp[[i]]$Strand=="+", match(paste(introns.ret.samp[[i]]$ENST, introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$End, sep = ":"), paste(all.exons$ENST, all.exons$Chromosome, all.exons$Start, sep = ":")),
                                               match(paste(introns.ret.samp[[i]]$ENST, introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start, sep = ":"), paste(all.exons$ENST, all.exons$Chromosome, all.exons$End, sep = ":"))),]
}


for (i in 1:length(introns.ret.samp)) {
  exons.retint.flank5ss[[i]]<-na.omit(exons.retint.flank5ss[[i]])
  exons.retint.flank3ss[[i]]<-na.omit(exons.retint.flank3ss[[i]])
  exons.retint.flank5ss[[i]]<-exons.retint.flank5ss[[i]][,c(1:3,7,5:6)] #replacing Code column with ENST column
  exons.retint.flank3ss[[i]]<-exons.retint.flank3ss[[i]][,c(1:3,7,5:6)]
}

## saving exonic coordinates to further use them in Linux environment
names(exons.retint.flank5ss)<-c("BlMo_Hm01", "BlMo_Hm03", "BlMo_Hm05", "BlMa_Hm03", "BlMa_Hm05")
for (i in 1:length(introns.ret.samp)) {
  write.table(as.data.frame(exons.retint.flank5ss[[i]]), file=paste0("/spectrum/GSCT/veronikap/DEEP/GCcount/", names(exons.retint.flank5ss)[i], ".retint.flankexons5ss.txt"), sep = "\t", col.names = F, row.names = F, quote = F)
}

names(exons.retint.flank3ss)<-c("BlMo_Hm01", "BlMo_Hm03", "BlMo_Hm05", "BlMa_Hm03", "BlMa_Hm05")
for (i in 1:length(introns.ret.samp)) {
  write.table(as.data.frame(exons.retint.flank3ss[[i]]), file=paste0("/spectrum/GSCT/veronikap/DEEP/GCcount/", names(exons.retint.flank3ss)[i], ".retint.flankexons3ss.txt"), sep = "\t", col.names = F, row.names = F, quote = F)
}



## determine GC contnet using bedtools and loading the output files back into R
GCcont.exons.flank5ss.files<-c("/spectrum/GSCT/veronikap/DEEP/GCcount/GC_BlMo_Hm01.retint.flankexons5ss.txt",
                               "/spectrum/GSCT/veronikap/DEEP/GCcount/GC_BlMo_Hm03.retint.flankexons5ss.txt",
                               "/spectrum/GSCT/veronikap/DEEP/GCcount/GC_BlMo_Hm05.retint.flankexons5ss.txt",
                               "/spectrum/GSCT/veronikap/DEEP/GCcount/GC_BlMa_Hm03.retint.flankexons5ss.txt",
                               "/spectrum/GSCT/veronikap/DEEP/GCcount/GC_BlMa_Hm05.retint.flankexons5ss.txt")

GCcont.exons.flank5ss.retint<-list()
for (i in 1:length(introns.ret.samp)) {
  GCcont.exons.flank5ss.retint[[i]]<-read.delim(GCcont.exons.flank5ss.files[i], sep = "\t", header = T)
}

for (i in 1:length(introns.ret.samp)) {
  introns.ret.samp[[i]]$Start_5ssexon<-GCcont.exons.flank5ss.retint[[i]]$X2_usercol[ifelse(introns.ret.samp[[i]]$Strand=="+", match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start, sep = ":"), paste0("chr",GCcont.exons.flank5ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank5ss.retint[[i]]$X3_usercol)),
                                                                                           match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$End, sep = ":"), paste0("chr",GCcont.exons.flank5ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank5ss.retint[[i]]$X2_usercol)))]
  introns.ret.samp[[i]]$End_5ssexon<- GCcont.exons.flank5ss.retint[[i]]$X3_usercol[ifelse(introns.ret.samp[[i]]$Strand=="+", match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start, sep = ":"), paste0("chr",GCcont.exons.flank5ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank5ss.retint[[i]]$X3_usercol)),
                                                                                          match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$End, sep = ":"), paste0("chr",GCcont.exons.flank5ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank5ss.retint[[i]]$X2_usercol)))]
  introns.ret.samp[[i]]$GCcont_5ssexon<-GCcont.exons.flank5ss.retint[[i]]$X8_pct_gc[ifelse(introns.ret.samp[[i]]$Strand=="+", match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start, sep = ":"), paste0("chr",GCcont.exons.flank5ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank5ss.retint[[i]]$X3_usercol)),
                                                                                           match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$End, sep = ":"), paste0("chr",GCcont.exons.flank5ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank5ss.retint[[i]]$X2_usercol)))]}

GCcont.exons.flank3ss.files<-c("/spectrum/GSCT/veronikap/DEEP/GCcount/GC_BlMo_Hm01.retint.flankexons3ss.txt",
                               "/spectrum/GSCT/veronikap/DEEP/GCcount/GC_BlMo_Hm03.retint.flankexons3ss.txt",
                               "/spectrum/GSCT/veronikap/DEEP/GCcount/GC_BlMo_Hm05.retint.flankexons3ss.txt",
                               "/spectrum/GSCT/veronikap/DEEP/GCcount/GC_BlMa_Hm03.retint.flankexons3ss.txt",
                               "/spectrum/GSCT/veronikap/DEEP/GCcount/GC_BlMa_Hm05.retint.flankexons3ss.txt")

GCcont.exons.flank3ss.retint<-list()
for (i in 1:length(introns.ret.samp)) {
  GCcont.exons.flank3ss.retint[[i]]<-read.delim(GCcont.exons.flank3ss.files[i], sep = "\t", header = T)
}

for (i in 1:length(introns.ret.samp)) {
  introns.ret.samp[[i]]$Start_3ssexon<-GCcont.exons.flank3ss.retint[[i]]$X2_usercol[ifelse(introns.ret.samp[[i]]$Strand=="+", match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$End, sep = ":"), paste0("chr",GCcont.exons.flank3ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank3ss.retint[[i]]$X2_usercol)),
                                                                                           match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start, sep = ":"), paste0("chr",GCcont.exons.flank3ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank3ss.retint[[i]]$X3_usercol)))]
  introns.ret.samp[[i]]$End_3ssexon<- GCcont.exons.flank3ss.retint[[i]]$X3_usercol[ifelse(introns.ret.samp[[i]]$Strand=="+", match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$End, sep = ":"), paste0("chr",GCcont.exons.flank3ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank3ss.retint[[i]]$X2_usercol)),
                                                                                          match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start, sep = ":"), paste0("chr",GCcont.exons.flank3ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank3ss.retint[[i]]$X3_usercol)))]
  introns.ret.samp[[i]]$GCcont_3ssexon<-GCcont.exons.flank3ss.retint[[i]]$X8_pct_gc[ifelse(introns.ret.samp[[i]]$Strand=="+", match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$End, sep = ":"), paste0("chr",GCcont.exons.flank3ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank3ss.retint[[i]]$X2_usercol)),
                                                                                           match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start, sep = ":"), paste0("chr",GCcont.exons.flank3ss.retint[[i]]$X.1_usercol,":",GCcont.exons.flank3ss.retint[[i]]$X3_usercol)))]}

for (i in 1:length(introns.ret.samp)) {
  introns.ret.samp[[i]]$Length_5ssexon<-introns.ret.samp[[i]]$End_5ssexon - introns.ret.samp[[i]]$Start_5ssexon
  introns.ret.samp[[i]]$Length_3ssexon<-introns.ret.samp[[i]]$End_3ssexon - introns.ret.samp[[i]]$Start_3ssexon
}


## loading PSI values for exons into R
PSI.exons.files<-c("/spectrum/GSCT/veronikap/DEEP/PSI/BlMo/Hm01_BlMo_exonic_parts.psi",
                   "/spectrum/GSCT/veronikap/DEEP/PSI/BlMo/Hm03_BlMo_exonic_parts.psi",
                   "/spectrum/GSCT/veronikap/DEEP/PSI/BlMo/Hm05_BlMo_exonic_parts.psi",
                   "/spectrum/GSCT/veronikap/DEEP/PSI/BlMa/Hm03_BlMa_exonic_parts.psi",
                   "/spectrum/GSCT/veronikap/DEEP/PSI/BlMa/Hm05_BlMa_exonic_parts.psi")

PSI.exons<-list()
for (i in 1:length(PSI.exons.files)) {
  PSI.exons[[i]]<-read.delim(PSI.exons.files[i], sep = "\t", header = T)
  PSI.exons[[i]]$chr<-sub("^", "chr", PSI.exons[[i]]$chr)
  PSI.exons[[i]]<-PSI.exons[[i]][order(PSI.exons[[i]]$chr, PSI.exons[[i]]$start),]
  PSI.exons[[i]]$split<-strsplit(as.character(PSI.exons[[i]]$exon_ID), split="/")
  PSI.exons[[i]]$ENST<-sapply(PSI.exons[[i]]$split, "[[", 1)
}

for (i in 1:length(introns.ret.samp)) {
  introns.ret.samp[[i]]$PSI_5ssexon<-PSI.exons[[i]]$PSI[match(paste(introns.ret.samp[[i]]$ENST, introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start_5ssexon, introns.ret.samp[[i]]$End_5ssexon, sep = ":"),
                                                              paste(PSI.exons[[i]]$ENST, PSI.exons[[i]]$chr, PSI.exons[[i]]$start, PSI.exons[[i]]$stop, sep=":"))]
  introns.ret.samp[[i]]$SplicedIn_5ssexon<-ifelse(introns.ret.samp[[i]]$PSI_5ssexon<0.9 | is.na(introns.ret.samp[[i]]$PSI_5ssexon), 0, 1)
  introns.ret.samp[[i]]$PSI_3ssexon<-PSI.exons[[i]]$PSI[match(paste(introns.ret.samp[[i]]$ENST, introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start_3ssexon, introns.ret.samp[[i]]$End_3ssexon, sep = ":"),
                                                              paste(PSI.exons[[i]]$ENST, PSI.exons[[i]]$chr, PSI.exons[[i]]$start, PSI.exons[[i]]$stop, sep=":"))]
  introns.ret.samp[[i]]$SplicedIn_3ssexon<-ifelse(introns.ret.samp[[i]]$PSI_3ssexon<0.9 | is.na(introns.ret.samp[[i]]$PSI_3ssexon), 0, 1)
}


## Loading Maximum Entropy scores for the splice sites
MaxEntScore.retint.5ss.files<-c("/spectrum/GSCT/veronikap/DEEP/MaxEntScan/BlMo_Hm01/BlMo_Hm01_retint_5ss.txt",
                                "/spectrum/GSCT/veronikap/DEEP/MaxEntScan/BlMo_Hm03/BlMo_Hm03_retint_5ss.txt",
                                "/spectrum/GSCT/veronikap/DEEP/MaxEntScan/BlMo_Hm05/BlMo_Hm05_retint_5ss.txt",
                                "/spectrum/GSCT/veronikap/DEEP/MaxEntScan/BlMa_Hm03/BlMa_Hm03_retint_5ss.txt",
                                "/spectrum/GSCT/veronikap/DEEP/MaxEntScan/BlMa_Hm05/BlMa_Hm05_retint_5ss.txt")

MaxEntScore.retint.5ss<-list()
for (i in 1:length(introns.ret.samp)) {
  MaxEntScore.retint.5ss[[i]]<-read.delim(MaxEntScore.retint.5ss.files[i], sep = "\t", header = F)
}

MaxEntScore.retint.3ss.files<-c("/spectrum/GSCT/veronikap/DEEP/MaxEntScan/BlMo_Hm01/BlMo_Hm01_retint_3ss.txt",
                                "/spectrum/GSCT/veronikap/DEEP/MaxEntScan/BlMo_Hm03/BlMo_Hm03_retint_3ss.txt",
                                "/spectrum/GSCT/veronikap/DEEP/MaxEntScan/BlMo_Hm05/BlMo_Hm05_retint_3ss.txt",
                                "/spectrum/GSCT/veronikap/DEEP/MaxEntScan/BlMa_Hm03/BlMa_Hm03_retint_3ss.txt",
                                "/spectrum/GSCT/veronikap/DEEP/MaxEntScan/BlMa_Hm05/BlMa_Hm05_retint_3ss.txt")

MaxEntScore.retint.3ss<-list()
for (i in 1:length(introns.ret.samp)) {
  MaxEntScore.retint.3ss[[i]]<-read.delim(MaxEntScore.retint.3ss.files[i], sep = "\t", header = F)
}

for (i in 1:length(introns.ret.samp)) {
  introns.ret.samp[[i]]$Start5ss<-ifelse(introns.ret.samp[[i]]$Strand=="+", introns.ret.samp[[i]]$Start-3, introns.ret.samp[[i]]$End-6)
  introns.ret.samp[[i]]$End5ss<-ifelse(introns.ret.samp[[i]]$Strand=="+", introns.ret.samp[[i]]$Start5ss+9, introns.ret.samp[[i]]$Start5ss+9)
  introns.ret.samp[[i]]$Start3ss<-ifelse(introns.ret.samp[[i]]$Strand=="+", introns.ret.samp[[i]]$End-20, introns.ret.samp[[i]]$Start-3)
  introns.ret.samp[[i]]$End3ss<-ifelse(introns.ret.samp[[i]]$Strand=="+", introns.ret.samp[[i]]$Start3ss+23, introns.ret.samp[[i]]$Start3ss+23)
  introns.ret.samp[[i]]$MaxEntScore5ss<-MaxEntScore.retint.5ss[[i]]$V7[match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start5ss, introns.ret.samp[[i]]$End5ss, sep=":"), paste0("chr",MaxEntScore.retint.5ss[[i]]$V1,":",MaxEntScore.retint.5ss[[i]]$V2,":",MaxEntScore.retint.5ss[[i]]$V3))]
  introns.ret.samp[[i]]$MaxEntScore3ss<-MaxEntScore.retint.3ss[[i]]$V7[match(paste(introns.ret.samp[[i]]$Chromosome, introns.ret.samp[[i]]$Start3ss, introns.ret.samp[[i]]$End3ss, sep=":"), paste0("chr",MaxEntScore.retint.3ss[[i]]$V1,":",MaxEntScore.retint.3ss[[i]]$V2,":",MaxEntScore.retint.3ss[[i]]$V3))]
  introns.ret.samp[[i]]$Strength5ss<-sapply(introns.ret.samp[[i]]$MaxEntScore5ss, function(x, na.rm = T) if (is.na(x)) "NA"
                                            else if (x<=5) "Weak"
                                            else if (x<=9) "Medium"
                                            else if (x<=15) "Strong")
  introns.ret.samp[[i]]$Strength3ss<-sapply(introns.ret.samp[[i]]$MaxEntScore3ss, function(x, na.rm = T) if (is.na(x)) "NA"
                                            else if (x<=5) "Weak"
                                            else if (x<=9) "Medium"
                                            else if (x<=15) "Strong")
}

 save(introns.ret.samp, file = "/home/veronikap/MONvsMAC/introns.ret.samp.RData")

 ## the same procedure is repeated for non-retained introns
